zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/changes/PartitionEnableExporterApplier.java

```diff
@@ -0,0 +1,144 @@
+    }
+
+    final MemberState member = currentClusterConfiguration.getMember(memberId);
+    final var partitionExistsInLocalMember = member.hasPartition(partitionId);
+
+    if (!partitionExistsInLocalMember) {
+      return Either.left(
+          new IllegalStateException(
+              String.format(
+                  "Expected to enable exporter, but the member '%s' does not have the partition '%s'",
+                  memberId, partitionId)));
+    }
+
+    final var exportersInPartition =
+        member.getPartition(partitionId).config().exporting().exporters();
+
+    if (initializeFrom.isPresent()) {
+      final String otherExporterId = initializeFrom.orElseThrow();
+      final var partitionHasOtherExporter = exportersInPartition.containsKey(otherExporterId);
+      if (!partitionHasOtherExporter) {
+        return Either.left(
+            new IllegalStateException(
+                String.format(
+                    "Expected to enable exporter and initialize from exporter '%s', but the partition '%s' does not have exporter '%s'",
+                    otherExporterId, partitionId, otherExporterId)));
+      } else {
+        final var otherExporter = exportersInPartition.get(otherExporterId);
+        if (otherExporter.state() == State.DISABLED) {
+          return Either.left(
+              new IllegalStateException(
+                  String.format(
+                      "Expected to enable exporter and initialize from exporter '%s', but the exporter '%s' is disabled",
+                      otherExporterId, otherExporterId)));
+        }
+      }
+    }
+
+    final var partitionHasExporter = exportersInPartition.containsKey(exporterId);
+
+    if (partitionHasExporter) {
+      // Increment by one when re-enabling the exporter
+      metadataVersionToUpdate = exportersInPartition.get(exporterId).metadataVersion() + 1;
+    } else {
+      metadataVersionToUpdate = 1;
+    }
+
+    return Either.right(memberState -> memberState);
+  }
+
+  @Override
+  public ActorFuture<UnaryOperator<MemberState>> applyOperation() {
+    final var result = new CompletableActorFuture<UnaryOperator<MemberState>>();
+
+    final ActorFuture<Void> enableFuture =
+        initializeFrom
+            .map(
+                otherExporterId ->
+                    partitionChangeExecutor.enableExporter(
+                        partitionId, exporterId, metadataVersionToUpdate, otherExporterId))
+            .orElse(
+                partitionChangeExecutor.enableExporter(
+                    partitionId, exporterId, metadataVersionToUpdate));
+
+    enableFuture.onComplete(
+        (nothing, error) -> {
+          if (error == null) {
+            result.complete(
+                memberState ->
+                    memberState.updatePartition(
+                        partitionId, partition -> partition.updateConfig(this::enableExporter)));
+          } else {
+            result.completeExceptionally(error);
+          }
+        });
```

Comment:
:upside_down_face: This would've been a great place to introduce some sort of `thenApply` to `ActorFuture`, which I think would be quite easy. The main problem however is that here we don't have an explicit executor, I notice we rely a lot on the executor-less `onComplete` method. I guess this is common in this module, but I don't know if this is a good habit to pick up, considering the comment for that code path is specifically it's only useful for tests :sweat_smile:

Reply:
We have `andThen` which is similar to `CompletableFuture#thenComponse`. But it is only recommended to use in specific situation as it has synchronization overhead and it also needs an executor. 

`ActorFuture#onComplete` has always been used in production code. It is not restricted to tests. `onComplete` with the executor must be used if we know that caller is not actor. So here it is perfectly acceptible to use the normal `onComplete`.

Reply:
Hm, here's the implementation:

```java
@Override
  public void onComplete(final BiConsumer<V, Throwable> consumer) {
    if (ActorThread.isCalledFromActorThread()) {
      final ActorControl actorControl = ActorControl.current();
      actorControl.runOnCompletion(this, consumer);
    } else {
      // We don't reject this because, this is useful for tests. But the warning is a reminder not
      // to use this in production code.
      Loggers.ACTOR_LOGGER.warn(
          "No executor provided for ActorFuture#onComplete callback."
              + " This could block the actor that completes the future."
              + " Use onComplete(consumer, executor) instead.");
      onComplete(consumer, Runnable::run);
    }
  }

```

This is what I basing it off of. And I think it makes sense in general, we want to usually control where things are executed, and having it run in the thread which completes the future can have unintended side effects :shrug:

Reply:
Regarding `andThen` and the recommendation, I was surprised by it. The comment that it creates intermediate future is not really a concern imo - this is exactly what `CompletableFuture` does, in fact (mostly because it needs to build its whole pipelining strategy). And then scheduling the callback on the executor is, well, also expected? I was a little curious about that recommendation in general :shrug: